### DISCUSSION TITLE
How does YALMIP deal with inequality constraints in SDPs?

### DISCUSSION THREAD

**Discussion by benebrue** [yalmip/YALMIP#1130] [General discussion] [answered]
Because of performance considerations, I was wondering about how YALMIP processes inequality constraints in SDPs.
The SDP in standard form as it's found in most of the classical literature often only contains equality constraints.
However, I've noticed that there exist some solvers such as [SDPNAL+](https://arxiv.org/abs/1710.10604) which are supported by YALMIP and claim they can natively deal with inequality constraints without requiring a reformulation to equality constraints using slack variables.

I wasn't able to find any infos on this on the YALMIP website, so some insights on how these inequalities are handled internally would be much appreciated!

**Comment by johanlofberg** [comment, marked as answer]
I don't know of any relevant solver which does not handle equalities and inequalities natively.

However, the way you (i.e. YALMP) interpret a model with equalities can make impact https://yalmip.github.io/tutorial/automaticdualization

However, that paper talks about semidefinite bound constraints which SDPNAL handles directly in the primal cone. This is not exploited via YALMIP. They would just lead to two additional SDP constraints, like any other constraints, on the dual side, or two slacks with associated cones and equalities on the primal if one would chose to dualize (which would be a bad idea as there would be a massive amount of equalities)

**Reply by johanlofberg** [reply]
Correction: that paper does not work with semidefinite bounds, it was just simple linear inequalities in the primal model. YALMIP will just use slacks if you chose to dualize this and interpret from primal side.
